coderd/database/databasefake/databasefake.go

```diff
@@ -871,6 +871,54 @@ func (q *fakeQuerier) GetAuthorizedWorkspaces(ctx context.Context, arg database.
 			}
 		}
 
+		if arg.HasAgent != "" {
+			build, err := q.GetLatestWorkspaceBuildByWorkspaceID(ctx, workspace.ID)
+			if err != nil {
+				return nil, xerrors.Errorf("get latest build: %w", err)
+			}
+
+			job, err := q.GetProvisionerJobByID(ctx, build.JobID)
+			if err != nil {
+				return nil, xerrors.Errorf("get provisioner job: %w", err)
+			}
+
+			workspaceResources, err := q.GetWorkspaceResourcesByJobID(ctx, job.ID)
+			if err != nil {
+				return nil, xerrors.Errorf("get workspace resources: %w", err)
+			}
+
+			var workspaceResourceIDs []uuid.UUID
+			for _, wr := range workspaceResources {
+				workspaceResourceIDs = append(workspaceResourceIDs, wr.ID)
+			}
+
+			workspaceAgents, err := q.GetWorkspaceAgentsByResourceIDs(ctx, workspaceResourceIDs)
+			if err != nil {
+				return nil, xerrors.Errorf("get workspace agents: %w", err)
+			}
+
+			var hasAgentMatched bool
+			for _, wa := range workspaceAgents {
+				switch arg.HasAgent {
+				case "connected":
+					hasAgentMatched = wa.LastConnectedAt.Valid
+				case "connecting":
+					hasAgentMatched = !wa.FirstConnectedAt.Valid
+				case "disconnected":
+					hasAgentMatched = (wa.DisconnectedAt.Valid && wa.DisconnectedAt.Time.After(wa.LastConnectedAt.Time)) ||
+						(wa.LastConnectedAt.Valid && wa.LastConnectedAt.Time.Add(time.Duration(arg.AgentInactiveDisconnectTimeout)*time.Second).Before(database.Now()))
+				case "timeout":
+					hasAgentMatched = !wa.FirstConnectedAt.Valid &&
+						wa.CreatedAt.Add(time.Duration(wa.ConnectionTimeoutSeconds)*time.Second).Before(database.Now())
+				}
+				break // only 1 agent is expected
```

Comment:
This comment confused me a bit at first, but then I noticed that we'll always only check one agent here. Should this be something like this instead?

```suggestion
				if hasAgentMatched {
					break // only 1 agent is required to match
				}
```

Reply:
I guess my question is, what happens when there are multiple agent, and with different statuses? Let's say we want workspaces with timed out agents. We could have one workspace with two agents where the first is connected and the second is timed out. In this case, I'd expect this workspace to show up in the search result but we'd only be checking the first one, i.e. connected.

Reply:
Let me refer to the SQL implementation. This is the order:

```
	ORDER BY
		build_number DESC,
		agent_last_connected_at ASC,
		agent_first_connected_at ASC
	LIMIT
		1
```

So, in practice, the statement will only consider that last build and look for the last connected agent. To break this assumption we would need a build job will multiple agents. Please correct me if I'm wrong, as this is the core part of this PR :)

Reply:
Ah right. So IMO it makes sense we're looking at the latest build, but we should probably consider all agents belonging to that build in the query. 🤔 It'd be strange if only the last agent affected it.

Reply:
Ok, I will try to improve the code. I admit that currently, it looks only at the last build and the most up-to-date agent.

Reply:
Refactored.